Use edx-ace for the password reset email

OmarIthawi opened the pull request:
I've pivoted to use edX ACE instead of manually rendering the HTML.

<kbd><img width="450" src="https://user-images.githubusercontent.com/645156/35188664-952aa728-fe42-11e7-82ac-acc9abf4dd46.png" /></kbd>

### TODO
 - [x] Get the Django email channel PR merged: https://github.com/edx/edx-ace/pull/25
    - [x] Bump the versions to include `edx-ace==0.1.7`
 - [ ] Test if locally one more time before it's merged.
 - [x] Nimisha to check if `site.domain` usage is correct.
 - [x] Nimisha to check if the `platform_name` override is correct.
 - [x] Spice up the HTML template for password reset.

### Reviewers
 - [x] A review from @nasthagiri 
 - [ ] Get it reviewed @grozdanowski for theming and Microsite friendliness

 ----

## Too Long Don't Read :point_down: 

> This is a leap of faith! I'm trying to get this merged after being on hold for a while!
> 
> ### Description
> Add in support for multipart emails (html) in the automatic messages that respond to user actions e.g: account creation, password change and password reset.
> 
> This PR:
> 
>  - Adds all the required templates to render the html messages (the base template is the one from the bulk_email app which ultimately is the one from the mailchimp templates)
>  - Overrides the save method in class PasswordResetFormNoActive to have HTML template.
> The bulk of this work was completed by @felipemontoya in #3374, and Sarina's #6681. 
> 
> I am taking over from them, to add in generic branding, get tests passing, and finish code review.
>

edx-webhook commented:
Thanks for the pull request, @OmarIthawi! I've created [OSPR-2009](https://openedx.atlassian.net/browse/OSPR-2009) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

nedbat commented:
@OmarIthawi I'll keep my fingers crossed that your leap is successful!  Ping me when this is ready for review.

OmarIthawi commented:
Thanks @nedbat!

OmarIthawi commented:
jenkins run all

OmarIthawi commented:
jenkins run quality

OmarIthawi commented:
Hi @nedbat this PR is mostly done and ready for review.

// cc: @felipemontoya

nasthagiri commented:
Please note that we're moving towards using [ACE](https://edx-ace.readthedocs.io/en/latest/) for our transactional emails in the platform.  You can see how it is used in [Schedules](https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/schedules/docs/README.rst) and in [Discussions](https://github.com/edx/edx-platform/tree/master/lms/djangoapps/discussion/templates/discussion/edx_ace/responsenotification/email).

OmarIthawi commented:
@nasthagiri Could you please take a look at the the latest changes? I've finished `DjangoEmailChannel`, and now `password_reset()` view to use ACE.

OmarIthawi commented:
jenkins run bokchoy

nasthagiri commented:
@OmarIthawi yes, looks good to me.
Would you want to make what you have already ready for merging - so we can merge those changes  as you work on the rest?

OmarIthawi commented:
Thanks for the suggestion @nasthagiri! Looks like a good idea to me.

I'll go ahead and start cleaning up. Will ping you when I'm done, hopefully very soon.

nasthagiri commented on openedx/core/djangoapps/ace_common/apps.py:
```diff
@@ -20,6 +20,7 @@ class AceCommonConfig(AppConfig):
                 SettingsType.AWS: {PluginSettings.RELATIVE_PATH: u'settings.aws'},
                 SettingsType.COMMON: {PluginSettings.RELATIVE_PATH: u'settings.common'},
                 SettingsType.DEVSTACK: {PluginSettings.RELATIVE_PATH: u'settings.common'},
+                SettingsType.TEST: {PluginSettings.RELATIVE_PATH: u'settings.common'},
```
I don't believe this line is needed since the `test` settings always import `common` settings.  See https://github.com/edx/edx-platform/blob/master/lms/envs/docs/README.rst

OmarIthawi replied:
Thanks for noting that @nasthagiri!

nasthagiri commented on openedx/core/djangoapps/ace_common/settings/common.py:
```diff
@@ -10,6 +10,9 @@ def plugin_settings(settings):
     settings.ACE_CHANNEL_SAILTHRU_API_SECRET = None
 
+    email_getter = 'openedx.core.djangoapps.site_configuration.helpers:get_email_from_address'
+    settings.ACE_CHANNEL_DJANGO_FROM_ADDRESS_GETTER = email_getter
```
Hooray!  Looks like you're tackling an issue that our White Label team is needing to address: [Customize From/Sender Emails for White Label in Sailthru (WL-1388)](https://openedx.atlassian.net/browse/WL-1388).
FYI @bill-filler

OmarIthawi replied:
Glad I'm helping!

bill-filler replied:
@OmarIthawi @nasthagiri 
excellent :)

nasthagiri commented on openedx/core/djangoapps/ace_common/templates/ace_common/edx_ace/common/base_body.html:
```diff
@@ -20,9 +20,10 @@
-{# Note {beacon_src} is not a template variable that is evaluated by the Django template engine. It is evaluated by #}
-{# Sailthru when the email is sent. Other email providers would need to replace this variable in the HTML as well. #}
-<img src="{beacon_src}" alt="" role="presentation" aria-hidden="true" />
+{# Note {beacon_html} is not a template variable that is evaluated by the Django template engine. #}
```
I **like** how you separated out the Sailthru-specific template values into their own Sailthru template files.  

Do you agree @mulby?

mulby replied:
I also like the separation, however, I wonder if we couldn't delegate some of this to the channel more directly?

That is, when we render the template, pass in the channel that it is being rendered for it, and allow the template to customize itself based on destination channel. I think this would clean up a bunch of stuff downstream.

For example, the beacon handling could look something like this:

```{{ destination_channel.beacon_html | safe }}```

Then, in the sailthru channel, we could simply implement an attribute that returns the appropriate HTML:

```
@property
def beacon_html(self):
    # Note {beacon_src} is not a template variable that is evaluated by the Django template engine. It is evaluated by ...
    return '<img src="{beacon_src}" alt="" role="presentation" aria-hidden="true" />'
```

Similarly, for the action links:

```
                         <!-- Actions -->
                         <td style="padding-bottom: 20px;">
                             {# Note that these variables are evaluated by Sailthru, not the Django template engine #}
                             {% for action_link in destination_channel.action_links %}
                             <p>
                                 <a href="{{ action_link.url }}" style="color: #005686">
                                     {# Note that the action link text must be translated before being inserted here #}
                                     <font color="#005686"><b>{{ action_link.text }}</b></font>
                                 </a>
                             </p>
                             {% endfor %}
                         </td>
```

nasthagiri replied:
@mulby Which delivery channel ultimately gets used is eventually determined by ACE in policy.py and delivery.py.  And beacon and action-links are channel-type specific.  For example, push notifications may not require them.

Does it make sense to push the `beacon_html` files into edx-ace instead of keeping them in edx-platform?  I don't imagine different IDAs needing their own versions.

And for the issue of having password-reset emails use Django-email instead of Sailthru, we can introduce a `Policy` implementation that requires Django-email over Sailthru for transactional-type emails.

mulby replied:
Another challenge we are going to have is that right now we only allow a single channel implementation for each type to be enabled at once. ACE will throw an error if there are multiple enabled email channels.

See this check: https://github.com/edx/edx-ace/blob/master/edx_ace/channel/__init__.py#L94

nasthagiri commented on openedx/core/djangoapps/user_api/accounts/api.py:
```diff
@@ -420,9 +420,11 @@ def request_password_change(email, is_secure):
     if form.is_valid():
         # Generate a single-use link for performing a password reset
         # and email it to the user.
+        from openedx.core.djangoapps.theming.helpers import get_current_request
```
Nit: Is it necessary for this to be an inline import?

OmarIthawi replied:
Not at all, just a left over!

OmarIthawi commented on common/djangoapps/student/forms.py:
```diff
@@ -47,41 +58,42 @@ def clean_email(self):
-            send_mail(subject, email, from_email, [user.email])
+                'request': request,
+                # TODO: This overrides what `get_base_template_context` proviedes, yet this make the tests passes
+                'platform_name': configuration_helpers.get_value('platform_name', settings.PLATFORM_NAME),
```
@nasthagiri Well, site name, platform name, base domain have too many ways to set in edx-platform. I'd love to know which are the "correct" ways.

OmarIthawi replied:
@nasthagiri TBH I have no idea, it would be great if you let me!

douglashall replied:
We should favor the capitalized form `configuration_helpers.get_value('PLATFORM_NAME', settings.PLATFORM_NAME)`.

nasthagiri replied:
@OmarIthawi Can you update this to request the capitalized form, as recommended by @douglashall?

OmarIthawi replied:
@douglashall @nasthagiri done!

OmarIthawi commented on common/djangoapps/student/forms.py:
```diff
@@ -47,41 +58,42 @@ def clean_email(self):
+                'platform_name': configuration_helpers.get_value('platform_name', settings.PLATFORM_NAME),
+                'reset_link': '{protocol}://{site}{link}'.format(
+                    protocol='https' if use_https else 'http',
+                    site=site.domain,
```
@nasthagiri Also this one (`site.domain`) could be revised, because I'm not completely sure it's correct.

nasthagiri replied:
@bill-filler I don't believe it's a blocker for this PR, but can you or someone from your team help us define and write the best practices for getting site-specific configuration?  See this question plus the TODO above.

bill-filler replied:
@nasthagiri 
Good idea. I will add a story to our backlog to write a best practices for getting site-specific configuration

OmarIthawi commented:
Thanks for the review @nasthagiri. I've addressed your latest comments. The PR is done from my end, except for the TODO items above.

Please let me know what do you think.

OmarIthawi commented:
jenkins run python

mulby commented on openedx/core/djangoapps/ace_common/templates/ace_common/edx_ace/common/base_body.html:
```diff
@@ -20,9 +20,10 @@
     {% block preview_text %}{% endblock %}
 </div>
 
-{# Note {beacon_src} is not a template variable that is evaluated by the Django template engine. It is evaluated by #}
```
We also need to make these changes to the red theme template

OmarIthawi commented:
Thanks @mulby for the review. There some nontrivial changes there! I'll see what I can do about them.

nasthagiri commented on common/djangoapps/student/forms.py:
```diff
@@ -47,41 +58,42 @@ def clean_email(self):
+                language=get_user_preference(user, LANGUAGE_KEY),
+                user_context=message_context,
+            )
+            ace.send(msg)
```
If we allow the message to specify the email channel type to use (as you suggested), then password-reset emails can continue to use Django-email, but through ACE.  ACE still provides other benefits - like a common framework for translations, white label override, Django templates, etc.

nasthagiri commented:
@OmarIthawi  Is it clear to you what remaining required work is left before merging this?

OmarIthawi commented:
Hi @nasthagiri I tried to contact @mulby via Open edX Slack but couldn't reach him. Copying the message here again:

> Hi Gabe
> This is Omar Al-Ithawi
> You've just reviewed a PR of mine: https://github.com/edx/edx-platform/pull/16545 (edited)
> There few non trivial changes need to be done on the PR
> I was wondering if we can jump into a call to discuss them
> I'd like to get this PR merged soon, since I have a lot of other planned opensource contributions
> But this one has been open since Nov 2017 (edited)
> 
> Omar Al-Ithawi [18:48]
> Just to give you a little bit of a background, usually OSPRs take a very long time because of discussion latency
> As a contributor this is a pain-point that I'm still unable to address (edited)
> While it's not a blocker in most of the cases
> As soon as there are architectural changes in the PR, it's an indicator for that PR may never get merged (edited)

OmarIthawi commented:
@mulby could you please take a look here?

OmarIthawi commented:
@mulby just a reminder about this PR.

nedbat commented:
@OmarIthawi Maybe we don't understand the state of this? The tests aren't passing yet.  What feedback are you looking for from @mulby ?

OmarIthawi commented:
Thanks @nedbat. The PR request has been refactored multiple times, so I was hesitant to fix the tests on something that could be refactored as well. 

Currently this PR is blocked by https://github.com/edx/edx-ace/pull/25 , which suffered from the same issue:

 1. Proposing a feature and making necessary changes to the code
 2. Fixing quality and tests issues
 3. Getting review feedback on architectural changes, back to point no. 1

I'm not saying that those changes are not necessary, however, this cycle is very expensive time-wise, and tend to be very taxing in terms of context switching because I only get the feedback later, when I've mostly dug deep into something else. 

If the PR was addressing a small bug, or a tweak, none of the above would have been a problem. This is why I introduce large PRs such as this one, only when absolutely necessary.

OmarIthawi commented:
@mulby this PR is ready for another go. I've addressed all the previous concerns. Unless Jenkins spots an issue, all the checks should be green by now.

// cc: @nasthagiri @nedbat

OmarIthawi commented on openedx/core/djangoapps/ace_common/settings/common.py:
```diff
@@ -9,6 +9,8 @@ def plugin_settings(settings):
     settings.ACE_CHANNEL_SAILTHRU_API_KEY = None
     settings.ACE_CHANNEL_SAILTHRU_API_SECRET = None
+    settings.ACE_CHANNEL_DEFAULT_EMAIL = 'django_email'
+    settings.ACE_CHANNEL_TRANSACTIONAL_EMAIL = 'django_email'
```
@mulby :+1:. However, it makes sense only in devstack, because, `django.core.mail` is the most helpful in tests.

OmarIthawi replied:
I've already done that. I think that the filebased email backend provides a nice HTML email debugging method.

OmarIthawi commented on openedx/core/djangoapps/ace_common/message_types.py:
```diff
@@ -0,0 +1,23 @@
+
+
+@attr.s(cmp=False)
+class SitesAwareMessageType(MessageType):
```
@mulby that's aboslutely right. It turns out overriding the `__init__` method is the best way to do such overriding. Redefining `options` was very problematic at best. Since `attr.s` classes don't support both inheritance and overriding at the same time in an easy way, `__init__` override seemed like the best option.

OmarIthawi commented on common/djangoapps/student/forms.py:
```diff
@@ -10,19 +10,32 @@
 
 
+class PasswordReset(SitesAwareMessageType):
+    options = {
```
@mulby see: https://github.com/edx/edx-platform/pull/16545#discussion_r181632376

OmarIthawi commented on lms/templates/registration/password_reset_email.html:
```diff
@@ -7,9 +7,7 @@
 {% trans "If you did not request this change, you can ignore this email." %}
```
To my surprise, yes. It was used for `password_change_request_handler` in the case of email not found. I've refactored that view to use edX ace as well.

OmarIthawi commented:
@mulby I've addressed all of the comments, in a single git commit. Should be good to go.

Please only merge after a manual test was done again (regardless of who performs that test).

OmarIthawi commented:
@mulby TravisCI is not doing very well. `caniusepython3` is having a trouble with the latest `pip`. Looks like we have some flaky dependencies despite the huge efforts on `pip-tools` :frowning_face:.

OmarIthawi commented:
jenkins run all

OmarIthawi commented on requirements/edx/base.in:
```diff
@@ -64,7 +64,7 @@ djangorestframework-jwt
 dogapi==1.2.1                       # Python bindings to Datadog's API, for metrics gathering
 enum34==1.1.6
-edx-ace
+-e git+https://github.com/appsembler/edx-ace@omar/django-email-channel#egg=edx-ace==0.1.6.omar
```
@mulby I'm sure you're aware of this hack. Which we should stop and use proper `pip-tools` method once the edX ACE PR is done.

OmarIthawi commented on common/djangoapps/student/forms.py:
```diff
@@ -47,41 +54,39 @@ def clean_email(self):
+                'platform_name': configuration_helpers.get_value('PLATFORM_NAME', settings.PLATFORM_NAME),
+                'reset_link': '{protocol}://{site}{link}'.format(
+                    protocol='https' if use_https else 'http',
+                    site=site.domain,
```
I should've written a note about it. I remember switching to `site.domain` because it was more reliable or something like that.

OmarIthawi replied:
Looks like both are being used interchanebly:

```
omar@oryx:edx-platform $ git grep 'site\.domain' | wc -l
59
omar@oryx:edx-platform $ git grep SITE_NAME | wc -l
96
```

mulby replied:
Hrmmm... in that case I'm inclined to leave it as it was - given how critical this email is to the operation of an installation I'm wary of changing this without fully understanding the implications.

OmarIthawi commented on openedx/core/djangoapps/ace_common/settings/common.py:
```diff
@@ -1,6 +1,6 @@
 def plugin_settings(settings):
     settings.ACE_ENABLED_CHANNELS = [
-        'file_email'
+        'django_email'
```
Should be be doable.

OmarIthawi replied:
Done

OmarIthawi commented:
@mulby I've reverted the hacks and will do a proper version bump once https://github.com/edx/edx-ace/pull/25 is merged.

mulby commented:
jenkins run all

OmarIthawi commented:
@mulby please note that the latest edX-ACE version (`0.1.7`) haven't yet been published on [pypi](https://pypi.org/project/edx-ace/#history), so this PR remains blocked.

// cc: @nedbat

OmarIthawi commented:
Looks like the branch gone stale. I'll resolve the conflicts shortly.

OmarIthawi commented:
jenkins run quality

OmarIthawi commented:
jenkins run python

OmarIthawi commented:
jenkins run bokchoy

OmarIthawi commented:
jenkins run bokchoy

OmarIthawi commented:
@mulby This PR should be ready. I'm still waiting for Jenkins' :heavy_check_mark:. Appreciate it if you can take another look.

OmarIthawi commented:
@nedbat @mulby The tests pass. Please take another look. It should be ready.

OmarIthawi commented:
Just a reminder about this PR: @nedbat @mulby @nasthagiri.

edx-webhook commented:
@OmarIthawi 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/16545&entry.1484655318=https://github.com/OmarIthawi&entry.752974735=2017-11-13+16:22&entry.1917517419=2018-05-10+15:31&entry.2133058324=Yes) so we can improve your experience in the future.

nasthagiri commented:
Congratulations @OmarIthawi 

![warnerarchive-classic-film-deep-in-my-heart-3oEduJ0MsePuOkzTtS](https://media2.giphy.com/media/3oEduJ0MsePuOkzTtS/giphy.gif)

OmarIthawi commented:
@mulby @nasthagiri Thank you for your efforts!

mulby commented:
Ok - I've completed the revert.

@OmarIthawi can you open a new pull request off of master with these changes in it? We can then hold that open until @bill-filler says it's ready to go.

OmarIthawi commented:
>@OmarIthawi can you open a new pull request off of master with these changes in it? We can then hold that open until @bill-filler says it's ready to go.

@mulby looks like I only saw your comment after you've opened the request. Thanks!
